actionpack/lib/action_controller/metal/renderers.rb

```diff
@@ -25,9 +42,16 @@ module Renderers
     # Default values are <tt>:json</tt>, <tt>:js</tt>, <tt>:xml</tt>.
     RENDERERS = Set.new
```

Comment:
A small update here, with respect to JSON API and discussion in https://github.com/rails/rails/pull/23712, that we're going to make a distinction between different renderers and rendering different media types.

Longer discourse on how this plays out:

## Different renderers for rendering different media types

e.g. JSON API has a different mime-type from `application/json`, so should have its own renderer, such that: 
- `respond_to do |format| format.jsonapi { render jsonapi: model } end` 
  - i.e. when `request.format == Mime[:jsonapi]`, then it will use the `Renderers::RENDERERS[:jsonapi]` and return `response.content_type = Mime[:jsonapi]`

It would be a *bug* for `render json:` to respond to a request with the JSON API media type and not render a JSON API document nor set the proper response content_type.

```ruby
  MEDIA_TYPE = 'application/vnd.api+json'
  HEADERS = {
    response: { 'CONTENT_TYPE' => MEDIA_TYPE },
    request:  { 'ACCEPT' => MEDIA_TYPE }
  }
 Mime::Type.register MEDIA_TYPE, :jsonapi
 media_type = Mime[:jsonapi]
 ActionDispatch::Http::Parameters::DEFAULT_PARSERS[media_type] = lambda do |body|
  data = JSON.parse(body)
  data = { :_json => data } unless data.is_a?(Hash)
  data.with_indifferent_access
end
ActionController::Renderers.add :jsonapi do |json, options|
  json = serialize_jsonapi(json, options).to_json(options) unless json.is_a?(String)
  self.content_type ||= media_type
  headers.merge! HEADERS[:response]
  self.response_body = json
end
```

## Configuring how a given renderer serializes/formats the model

e.g. the JSON renderer is defined as 

```ruby
    ActionController::Renders.add :json do |json, options|
      json = json.to_json(options) unless json.kind_of?(String)

      if options[:callback].present?
        if content_type.nil? || content_type == Mime[:json]
          self.content_type = Mime[:js]
        end

        "/**/#{options[:callback]}(#{json})"
      else
        self.content_type ||= Mime[:json]
        json
      end
    end
```

But I want to customize how my `json` object is serialized, and don't want to change the behavior of `to_json` in my model.

For example, in development I want to pretty-print my JSON responses. 

I should only need to change 
  - `json.to_json(options)` to 
  - `JSON.pretty_generate(json.as_json(options), options)`, 
but the current interface requires me to:

```diff
+  ActionController::Renderers.remove :json
  ActionController::Renders.add :json do |json, options|
-      json = json.to_json(options) unless json.kind_of?(String)
+     unless json.is_a?(String)
+       json = json.as_json(options) if json.respond_to?(:as_json)
+       json = JSON.pretty_generate(json, options)
+     end

      if options[:callback].present?
        if content_type.nil? || content_type == Mime[:json]
          self.content_type = Mime[:js]
        end

        "/**/#{options[:callback]}(#{json})"
      else
        self.content_type ||= Mime[:json]
        json
      end
    end
```

This seems overkill when we're not changing the `response.content_type = Mime[:json]` (aka format) nor any other 'renderer' behavior.  We really just want to hook into the RHS of `response.response_body =`.

The [original implementation of ActiveModelSerializers back in 2011](https://github.com/rails/rails/commit/8ff7693a8dc61f43fc4eaf72ed24d3b8699191fe) was essentially at the same touch point: `json = json.active_model_serializer.new(json, serialization_scope) if json.respond_to?(:active_model_serializer)`.

### Updated Proposal

Over in https://github.com/rails/rails/pull/21496#discussion_r53316942 I just noticed that @dhh posted an interface that I really like.

```ruby
class ApplicationController < ActionController::Base
  # Takes a hash of form { 'renderer format' => 'callable that returns the response body' }
  serializing json: ->(object, options) { ActiveModel::SerializableResource.new(model, options) }
```

The Renderer would call something like
-  `serializing[:json].call(object, options)` 
- that defaults to `serializers[:json] = ->(object, options) { object.kind_of?(String) ? object : json.to_json(options) }`, 
- but can be customized and inherited by controllers:



I'm guessing that

> Here's a simple proposal for cutting down on the specialness of the configuration as much as possible

means that there would be a `class_attribute :serializing` and that the diff should be essentially:

```diff
+  class_attribute :serializing, instance_predicate: false
+  self.serializing = Hash.new do |h, format| 
+    fail ActionController::MissingSerializer, "There is no '#{format}' serializer.\n" <<
+                "Known serializers are #{serializing.keys}"
+  end
+  serializing[:json] = ->(object, options) { object.kind_of?(String) ? object : json.to_json(options) }
  ActionController::Renders.add :json do |json, options|
-      json = json.to_json(options) unless json.kind_of?(String)
+     json = serializing[:json].call(json, options)

      if options[:callback].present?
        if content_type.nil? || content_type == Mime[:json]
          self.content_type = Mime[:js]
        end

        "/**/#{options[:callback]}(#{json})"
      else
        self.content_type ||= Mime[:json]
        json
      end
    end
```